syntax = "proto3";

import "google/protobuf/empty.proto";
import "kv.proto";

package temp;

option optimize_for = LITE_RUNTIME;

service Test
{
  rpc RunOps(stream OpIn) returns (stream OpOut) {}

  // Subscribe to updates for a single key
  rpc Sub(externalexecutor.protobuf.KVKey)
      returns (stream externalexecutor.protobuf.KVKeyValue)
  {
  }

  // Publishes updates for a single key
  rpc Pub(externalexecutor.protobuf.KVKeyValue) returns (google.protobuf.Empty)
  {
  }

  rpc Stream(google.protobuf.Empty)
      returns (stream externalexecutor.protobuf.KVKeyValue)
  {
  }
}

message OpIn
{
  oneof op
  {
    EchoOp echo = 1;
    ReverseOp reverse = 2;
    TruncateOp truncate = 3;
  }
}

message EchoOp { string body = 1; }
message ReverseOp { string body = 1; }
message TruncateOp
{
  string body = 1;
  uint32 start = 2;
  uint32 end = 3;
}

message OpOut
{
  oneof result
  {
    EchoResult echoed = 1;
    ReverseResult reversed = 2;
    TruncateResult truncated = 3;
  }
}

message EchoResult { string body = 1; }
message ReverseResult { string body = 1; }
message TruncateResult { string body = 1; }